### PR TITLE
[SPARK-11724] [SQL] Change casting between int and timestamp to consistently treat int in seconds.

### PR DESCRIPTION
**Pull request by nongli**
Hive has since changed this behavior as well. https://issues.apache.org/jira/browse/HIVE-3454

### CODE REVIEW

**Comment by cloud-fan** [comment]
Good catch! The currently behaviour looks wierd to me, how does hive handle this? cc @yhuai

**Comment by nongli** [comment]
> Good catch! The currently behaviour looks wierd to me, how does hive handle this? cc @yhuai

It looks like Hive fixed this in 1.2. https://issues.apache.org/jira/browse/HIVE-3454

**Comment by yhuai** [comment]
Seems we need to update those golden files. I believe we can set `HIVE_HOME` to hive 1.2.1's dir and delete those golden files. Then, when we run those tests again, we will first let Hive generate results.

**Comment by yhuai** [comment]
Seems Hive is still wrong when the data type is int. See
```
hive> SELECT CAST(CAST(-1200 AS TIMESTAMP) AS INT);
OK
-2
Time taken: 0.047 seconds, Fetched: 1 row(s)
hive> SELECT CAST(CAST(-1200.0 AS TIMESTAMP) AS INT);
OK
-1200
Time taken: 0.046 seconds, Fetched: 1 row(s)
```
Let's delete those golden files and create our own test (using `test(...) {...}` instead of using `createQueryTest(..., ...)`).

**Comment by yhuai** [comment]
test this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -248,12 +249,17 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
       |IF(TRUE, CAST(NULL AS BINARY), CAST("1" AS BINARY)) AS COL18,
       |IF(FALSE, CAST(NULL AS DATE), CAST("1970-01-01" AS DATE)) AS COL19,
       |IF(TRUE, CAST(NULL AS DATE), CAST("1970-01-01" AS DATE)) AS COL20,
-      |IF(FALSE, CAST(NULL AS TIMESTAMP), CAST(1 AS TIMESTAMP)) AS COL21,
-      |IF(TRUE, CAST(NULL AS TIMESTAMP), CAST(1 AS TIMESTAMP)) AS COL22,
-      |IF(FALSE, CAST(NULL AS DECIMAL), CAST(1 AS DECIMAL)) AS COL23,
-      |IF(TRUE, CAST(NULL AS DECIMAL), CAST(1 AS DECIMAL)) AS COL24
+      |IF(TRUE, CAST(NULL AS TIMESTAMP), CAST(1 AS TIMESTAMP)) AS COL21,
+      |IF(FALSE, CAST(NULL AS DECIMAL), CAST(1 AS DECIMAL)) AS COL22,
+      |IF(TRUE, CAST(NULL AS DECIMAL), CAST(1 AS DECIMAL)) AS COL23
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: Seems we do not include the new golden file? I can generate it tomorrow.

**Comment by yhuai** [comment]
LGTM. Thank you for the fix! Merging it to master and branch 1.6.

**Pull request closed without merging**